### ISSUE TITLE
[React] Integrating with Library Input Components

### ISSUE THREAD

**Issue by eric-horodyski** [uNmAnNeR/imaskjs#429]
I'm having some trouble integrating IMask with the [Ionic Framework library](https://ionicframework.com/react). Like several other component libraries, the Ionic Framework provides its own input components -- nesting the inner HTML `<input>` tag. 

For reference, I am using the core IMask package `imask` instead of the React plugin `react-imask`. I was not able to make it very far with the `react-imask` library due to typing issues (and had to put this project away). A colleague of mine made headway with the core library and shared the code with me, which I'm now able to pick back up. 

While trying to create a custom component (a masked "currency input" component), I noticed that IMask is having several synchronization issues: 

1. If a character is entered that does not match the RegEx, a warning is thrown.
2. When introducing a radix (for the cents portion of the field), a warning is thrown.

In both cases, the warning is the following:

```
Element value was changed outside of mask. Syncronize mask using `mask.updateValue()` to work properly.
```

Thinking through this, I wonder if the issue is one of the following:

* React state (which binds the value to the input field) doesn't run in-tandem with IMask, or;
* Although IMask is bound to the inner `<input>` tag of the `IonInput` component via a `ref`, the "outer" portion of the `IonInput` component conflicts with IMask.

I have the sample application uploaded to the following repository: https://github.com/ehorodyski-ionic/fw-react-input-masking . The behavior can be observed when pulling the repo down, installing Node modules, then:

1. Running `npm run start` to launch the application in a browser (I was using Chrome).
2. Launch the browser's Development Tools
2. Input characters in the "Today's Payment" input field - either introduce a period (.) or a non-digit character.
3. Observe the warnings logged to the console of the Development Tools.

The relevant portion of code can be found in [`CurrencyInput.tsx`](https://github.com/ehorodyski-ionic/fw-react-input-masking/blob/master/src/components/CurrencyInput.tsx). The component is used in [`Home.tsx` LN47](https://github.com/ehorodyski-ionic/fw-react-input-masking/blob/master/src/pages/Home.tsx#L47). 

If it's easier for you to scan through the component code, I've copy and pasted it below additionally. 

Appreciate the help!

**CurrencyInput.tsx**:
```TSX
import React, { useCallback, useRef, useState } from 'react';
import { IonInput } from '@ionic/react';
import IMask from 'imask';

interface CurrencyInputProps {
  value: any;
  onChange: (val: any) => any;
}

const CurrencyInput: React.FC<CurrencyInputProps> = ({ value, onChange }) => {
  const maskRef = useRef<IMask.InputMask<any> | null>(null);

  const inputCallback = useCallback(
    async (input: HTMLIonInputElement) => {
      if (!input) return;
      const nativeInput = await input.getInputElement();

      const mask = IMask(nativeInput, {
        mask: '$num',
        lazy: true,
        blocks: {
          num: {
            mask: Number,
            thousandsSeparator: ',',
            scale: 2,
          },
        },
      })
        .on('accept', (e: any) => {
          onChange(mask.value);
        })
        .on('complete', (e: any) => {
          onChange(mask.value);
        });

      maskRef.current = mask;
    },
    [onChange],
  );

  return <IonInput value={value} ref={inputCallback} />;
};
export default CurrencyInput;
```

**Comment by uNmAnNeR** [comment]
hi!

> React state (which binds the value to the input field) doesn't run in-tandem with IMask, or;

yes, exactly. by default imask does not prevent `input` event while handling, so you probably need somehow to unbind default event handler (or prevent input event manually) and just use `accept` trigger from mask. Note that events from imask is synchronous (which is much easy to handle).

Second thing is that you have to syncronize mask state with your app state in both ways. Now you are only updating app state on input, but same needs to be done wiseversa.

i think at the end you will implement quite same logic as already done in react plugin, so may be you just can pick it up in some way https://github.com/uNmAnNeR/imaskjs/blob/master/packages/react-imask/src/mixin.js
just throw away logic with handling `unmask` prop if you don't need it

**Comment by ehorodyski-ionic** [comment]
Thanks for the response @uNmAnNeR! The difficulty in using the React imask plugin stems from an inability to declare the proper TypeScript types, if I recall correctly. 

While I could mix-and-match TypeScript with Javascript files, our framework is built to utilize TypeScript as a first-class citizen; I'd prefer not to drop TypeScript from any code files. Using TypeScript with React, linters are _very_ strict, making it hard to derive the types needed to work with `react-imask`

Would it be possible to obtain an example of using the `react-imask` within a TypeScript environment? Standard `<input>` element would work; my struggle was finding out how to create the types themselves to pass TypeScript compilation.

**Comment by uNmAnNeR** [comment]
should be fixed in 83f8ce8
